### PR TITLE
DimmerItems were not converting DecimalTypes correctly, also group it…

### LINKED ISSUE #1476: Group:Dimmer:AVG reporting DecimalType state, not PercentType
So if I have the items
```
Group:Dimmer:AVG DIMMER_GROUP
Dimmer DimmedLight (DIMMER_GROUP)
Dimmer DimmedLight2 (DIMMER_GROUP)
```
I would expect the value of the group to be the AVG of the child item states, in PercentType format.  So if we have one dimmer at 0, and another at 100 the state should be 50.  Instead this  is reported as 0.50000, like so:
```
{
  "members": [
    {
      "link": "http://localhost:8080/rest/items/DimmedLight",
      "state": "0",
      "stateDescription": {
        "pattern": "%d %%",
        "readOnly": false,
        "options": []
      },
      "type": "DimmerItem",
      "name": "DimmedLight",
      "label": "Dimmer",
      "category": "slider",
      "tags": [],
      "groupNames": [
        "DIMMER_GROUP"
      ]
    },
    {
      "link": "http://localhost:8080/rest/items/DimmedLight2",
      "state": "100",
      "stateDescription": {
        "pattern": "%d %%",
        "readOnly": false,
        "options": []
      },
      "type": "DimmerItem",
      "name": "DimmedLight2",
      "label": "Dimmer",
      "category": "slider",
      "tags": [],
      "groupNames": [
        "DIMMER_GROUP"
      ]
    }
  ],
  "link": "http://localhost:8080/rest/items/DIMMER_GROUP",
  "state": "0.50000000",
  "stateDescription": {
    "pattern": "(%d)",
    "readOnly": false,
    "options": []
  },
  "type": "GroupItem",
  "name": "DIMMER_GROUP",
  "label": "dimmer group",
  "tags": [],
  "groupNames": []
}
```

One solution is to change the AVG function here https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java#L238

so instead of 
```
for (Item item : items) {
                    DecimalType itemState = (DecimalType) item.getStateAs(DecimalType.class);
                    if (itemState != null) {
                        sum = sum.add(itemState.toBigDecimal());
                        count++;
                    }
                }
```

We might do something like
```
 for (Item item : items) {
                    DecimalType itemState;
                    if (item.getAcceptedDataTypes().contains(PercentType.class)) {
                        itemState = (DecimalType) item.getStateAs(PercentType.class);
                    } else {
                        itemState = (DecimalType) item.getStateAs(DecimalType.class);
                    }
                    if (itemState != null) {
                        sum = sum.add(itemState.toBigDecimal());
                        count++;
                    }
                }
```

This would of course need to be done in other arithmetic functions as well.  I wanted to get feedback before submitting a PR for this.  Thanks.

### PR DESCRIPTION
**Pull request by digitaldan**
…ems where not using the their base item when setting state.  This fixes #1476

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

### CODE REVIEW

**Comment by digitaldan** [comment]
This is related to closed PR #1486 .  I will make a few comments to the PR on what changed and why.   Also there were some formatting changes done automatically by eclipse, these were unintentional of course.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 1]
```diff
@@ -263,9 +263,15 @@ public void stateUpdated(Item item, State state) {
         }
     }
 
+    @Override
     public void setState(State state) {
         State oldState = this.state;
-        this.state = state;
+        if (baseItem != null) {
+            baseItem.setState(state);
+            this.state = baseItem.getState();
```
Comment (digitaldan) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Here we set the state of the base item, which should be responsible for converting the type to the right format, then use that state for the group item.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/DimmerItem.java [hunk 2]
```diff
@@ -79,6 +79,8 @@ public void setState(State state) {
             super.setState(PercentType.ZERO);
         } else if (state == OnOffType.ON) {
             super.setState(PercentType.HUNDRED);
+        } else if (state.getClass() == DecimalType.class) {
+            super.setState(new PercentType(((DecimalType) state).toBigDecimal().multiply(new BigDecimal(100))));
```
Comment (digitaldan) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/DimmerItem.java]: When setting the state, do a conversion from DecimalType to PercentType, what's interesting is that on line 101 below the class was already doing a conversion the other way when getting the state.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/DimmerItem.java [hunk 3]
```diff
@@ -96,8 +98,8 @@ public State getStateAs(Class<? extends State> typeClass) {
             return state.equals(PercentType.ZERO) ? OnOffType.OFF : OnOffType.ON;
         } else if (typeClass == DecimalType.class) {
             if (state instanceof PercentType) {
-                return new DecimalType(((PercentType) state).toBigDecimal().divide(new BigDecimal(100), 8,
-                        RoundingMode.UP));
+                return new DecimalType(
```
Comment (digitaldan) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/DimmerItem.java]: This was a formatting change, but does highlight we were already converting between decimal and percent types in one direction.

**Comment by digitaldan** [comment]
@kai usually I would not add so many comments to such a small PR, but I am well aware of the risk of changing core library items.  I searched through the OH1 and OH2 code base looking to see if this could break anything, but since nothing actually manipulates a DimmerItem directly outside of the core api, I am less worried about it.

**Comment by maggu2810** [comment]
Your are right, in getStateAs function there is already code to handle "state=PercentType, as=DecimalType" and "state=DecimalType and as=PercentType".
As accepted data types is percent type and not decimal type I thing it is okay, to convert a decimal type on setState already to a percent type.

**Comment by kaikreuzer** [comment]
Thanks @digitaldan, lgtm - let's hope that it indeed won't have any unexpected side effects :-)

**Pull request merged**